dynamic_role.go

```diff
@@ -0,0 +1,61 @@
+package openldap
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"time"
+
+	"github.com/hashicorp/vault/sdk/logical"
+)
+
+type dynamicRole struct {
+	// required fields
+	Name         string `json:"name"          mapstructure:"name"`
+	CreationLDIF string `json:"creation_ldif" mapstructure:"creation_ldif"`
+
+	// optional fields
+	UsernameTemplate string        `json:"username_template,omitempty" mapstructure:"username_template"`
+	DefaultTTL       time.Duration `json:"default_ttl,omitempty"       mapstructure:"default_ttl"`
+	MaxTTL           time.Duration `json:"max_ttl,omitempty"           mapstructure:"max_ttl"`
+}
+
+func retrieveDynamicRole(ctx context.Context, s logical.Storage, roleName string) (*dynamicRole, error) {
+	entry, err := s.Get(ctx, path.Join(dynamicRolePath, roleName))
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	var result *dynamicRole
```

Comment:
I think this should this be `dynamicRole` since you're sending in `&result` below. No allocation is occurring here (though I think DecodeJSON _might_ do some magic allocation, I don't like relying on that).

Reply:
The reason for a pointer here isn't for memory allocation, it's so the ownership of the role (and thus the ability to change the object) is passed to the caller. That being said, I changed `&result` to `result` in the `DecodeJSON` call since it's already a pointer.

Reply:
`result` is still a nil pointer though, which will cause json unmarshalling to panic. So we need either:

```
var result = new(dynamicRole)
```

or

```
var result dynamicRole
...DecodeJSON(&result)...
```

Reply:
You are correct, I've updated accordingly. I'm very confused though since this was working previously and I can't figure out why it was working instead of erroring...